default/AI/ShipDesignAI.py

```diff
@@ -1010,6 +1023,13 @@ def optimize_design(self, additional_parts=[], additional_hulls=[],
             planet = universe.getPlanet(pid)
             self.pid = pid
             self.update_species(planet.speciesName)
+
+            relevant_techs = []
+            if WEAPONS & self.useful_part_classes:
+                relevant_techs = [tech for tech in AIDependencies.WEAPON_UPGRADE_TECHS if tech_is_complete(tech)]
```

Comment:
Dilvish prefer list comprehensions, I prefer this `filter(tech_is_complete, AIDependencies.WEAPON_UPGRADE_TECHS)`